### DISCUSSION TITLE
How to make Github packages to the public

### DISCUSSION THREAD

**Discussion by hantsy** [community/community#25629] [Packages]
<p>I have created a simple maven archetype project : <a href="https://github.com/hantsy/maven-archetype-jakartaee9" class="inline-onebox" rel="noopener nofollow ugc">GitHub - hantsy/maven-archetype-jakartaee9: Maven archetype for Jakarta EE 9</a></p>
<p>And publish it via Github packages.</p>
<p>But read the docs from Github Packages, it requires authentication to install it.</p>
<ol>
<li>Is it possible to make the repository public to access(without an auth token)?</li>
<li>Is there an approach to sync this repository to the Maven Central repository?</li>
</ol>

**Comment by jcansdale** [comment, marked as answer]
<p>Hi <a class="mention" href="/hantsy">@hantsy</a>,</p>
<aside class="quote no-group" data-username="hantsy" data-post="1" data-topic="171321">
<div class="title">
<div class="quote-controls"></div>
 hantsy:</div>
<blockquote>
<ul>
<li>Is it possible to make the repository public to access(without an auth token)?</li>
</ul>
</blockquote>
</aside>
<p>At the moment a token is always required, but it is possible to encode/embed the token in your .pom file.</p>
<p>Here is how you would do it:</p>
<ol>
<li>Create a PAT with just the <code>read:packages</code> scope</li>
<li>Execute <code>$ docker run ghcr.io/jcansdale/gpr encode &lt;PAT&gt;</code>
</li>
<li>This will generate a <code>repositories</code> element you can use in your .pom file</li>
</ol>
<p>The <code>repositories</code> element will look something like this:</p>
<pre><code class="lang-auto">&lt;repositories&gt;
  &lt;repository&gt;
    &lt;id&gt;github-public&lt;/id&gt;
    &lt;url&gt;https://public:&amp;#102;xxxxxxx@maven.pkg.github.com/&lt;OWNER&gt;/*&lt;/url&gt;
  &lt;/repository&gt;
&lt;/repositories&gt;
</code></pre>
<p>Replace <code>&lt;OWNER&gt;</code> with your user or org name.</p>
<p>Note, the <code>read:packages</code> token will have access to your public and private packages. If you have any private packages you need to project, you might want to create a <a href="https://docs.github.com/en/developers/overview/managing-deploy-keys#machine-users" rel="noopener nofollow ugc">machine-user</a> account.</p>
<aside class="quote no-group" data-username="hantsy" data-post="1" data-topic="171321">
<div class="title">
<div class="quote-controls"></div>
 hantsy:</div>
<blockquote>
<ul>
<li>Is there an approach to sync this repository to the Maven Central repository?</li>
</ul>
</blockquote>
</aside>
<p>There isn’t currently a way to automatically sync with Maven Central.</p>
<p>I hope that helps!</p>

**Comment by hantsy** [comment]
<p>Thanks <a class="mention" href="/jcansdale">@jcansdale</a>, it is very helpful. I hope it can be added in the official docs.</p>

**Reply by hantsy** [reply]
@jcansdale  This solution seems not work now.

**Comment by mattwelke** [comment]
<p>I like this workaround because it shows how anyone can set up their Maven or Gradle projects to get access to packages published to GitHub Packages in public GitHub repository. It should be added to the docs because I came here about to report the issue. I didn’t realize that using a PAT with read packages scope was the recommended way. It would have saved me some time to just have those instructions in the docs (<a href="https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-apache-maven-registry" class="inline-onebox" rel="noopener nofollow ugc">Working with the Apache Maven registry - GitHub Docs</a>).</p>
<p>Right now, the docs hint at this by saying:</p>
<blockquote>
<p>You must use a personal access token with the appropriate scopes to publish and install packages in GitHub Packages.</p>
</blockquote>
<p>and</p>
<blockquote>
<p>You can authenticate to GitHub Packages with Apache Maven by editing your <em>~/.m2/settings.xml</em> file to include your personal access token.</p>
</blockquote>
<p>The problem is that as I read this quickstart, not knowing about GitHub Packages in detail, I’m thinking about my previous experience with PATs on GitHub, and I remember that PATs represent my GitHub account, so I’m thinking one would only need a PAT to install a Maven package from a private GitHub repository. I didn’t realize the PATs were also needed for public GitHub repositories.</p>
<p>Also, the docs could use some more examples on how to use the packages once they’re published to GitHub Packages. Being somewhat new to Java, I wasn’t familiar with how to configure my <code>build.gradle</code> file to download from more locations than just Maven Central. I eventually figured out I had to replace <code>repositories</code> in my <code>build.gradle</code> file with:</p>
<pre><code class="lang-auto">repositories {
    mavenCentral()
    maven {
        url "https://maven.pkg.github.com/mattwelke/*"
        credentials {
            username "mattwelke"
            password "&lt;redacted&gt;"
        }
    }
}
</code></pre>
<p>(and then, later, once I got it all working, I’d auth with the PAT more securely than hard coding it in the <code>build.gradle</code> file)</p>

**Comment by elect86** [comment]
I have another alternative, [magik](https://github.com/elect86/magik)

**Reply by gilberto-009199** [reply]
kkk
